src/compiler/file_uploader/index.ts

```diff
@@ -0,0 +1,187 @@
+        getCanisterId(canisterName),
+        getWebServerPort(),
+        getIdentityName()
+    );
+
+    const chunkSize = 2_000_000; // The current message limit is about 2 MiB
+
+    for (const [srcPath, destPath] of paths) {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await upload(srcPath, destPath, chunkSize, actor);
+    }
+}
+
+async function upload(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    if (!existsSync(srcPath)) {
+        throw new Error(`${srcPath} does not exist`);
+    }
+
+    const stats = await stat(srcPath);
+    if (stats.isDirectory()) {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await uploadDirectory(srcPath, destPath, chunkSize, actor);
+    } else {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await uploadFile(srcPath, destPath, chunkSize, actor);
+    }
+}
+
+async function uploadDirectory(
+    srcDir: string,
+    destDir: string,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    try {
+        const contents = await readdir(srcDir);
+        for (const name of contents) {
+            const srcPath = join(srcDir, name);
+            const destPath = join(destDir, name);
+            // Await each upload so the canister doesn't get overwhelmed by requests
+            await upload(srcPath, destPath, chunkSize, actor);
+        }
+    } catch (error) {
+        console.error(`Error reading directory: ${error}`);
+    }
+}
+
+async function uploadFile(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    console.info(`uploadFile: Uploading ${srcPath} to ${destPath}`);
+    const uploadStartTime = process.hrtime.bigint();
+    const fileSize = (await stat(srcPath)).size;
+    for (let i = 0; i < fileSize; i += chunkSize) {
+        const fileStream = createReadStream(srcPath, {
+            start: i,
+            end: i + chunkSize - 1,
+            highWaterMark: chunkSize
+        });
+
+        let startIndex = i;
+        for await (const data of fileStream) {
+            await throttle();
+            console.info(
+                `uploadFile: ${srcPath} | ${bytesToHumanReadable(
+                    i + data.length
+                )} of ${bytesToHumanReadable(fileSize)}`
+            );
+            // Don't await here! Awaiting the agent will result in about a 4x increase in upload time.
+            // The above throttling is sufficient to manage the speed of uploads
+            actor
+                .upload_file_chunk(destPath, uploadStartTime, i, data, fileSize)
+                .catch((error) => {
+                    console.error(error);
+                });
+            startIndex += data.length;
+        }
+    }
+    console.info(`uploadFile: finished ${srcPath}`);
+}
+
+function bytesToHumanReadable(sizeInBytes: number): string {
```

Comment:
Redo this with reduce to get rid of the mutation

Reply:
done

Reply:
Although I will standby what I said about this not being as simple as the other one